### PR TITLE
LiteX I2S support

### PR DESCRIPTION
**Pull request by mateusz-holenko**
This adds support for the LiteX I2S core (https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/i2s.py): a driver and a sample demo application.

### CODE REVIEW

File: drivers/i2s/Kconfig.litex [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	int "RX queue length"
+	default 500
+
+config I2S_LITEX_TX_BLOCK_COUNT
+	int "TX queue length"
+	default 500
+
+config I2S_LITEX_CHANNELS_CONCATENATED
+	bool "Channels placed without padding in fifo"
+	default False
```
Comment (nzmichaelh) [inline drivers/i2s/Kconfig.litex:25]: False or n?

File: drivers/i2s/Kconfig.litex [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+config I2S_LITEX_DATA_BIG_ENDIAN
+	bool "Received data will be stored as big endian"
+	default False
```
Comment (nzmichaelh) [inline drivers/i2s/Kconfig.litex:29]: Ditto

File: drivers/i2s/i2s_litex.c [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+#include "i2s_litex.h"
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(i2s_litex);
+#define DEV_CFG(dev) ((struct i2s_litex_cfg *const)(dev)->config_info)
+#define DEV_DATA(dev) ((struct i2s_litex_data *const)(dev)->driver_data)
+
+#define MODULO_INC(val, max)                                                   \
+	{                                                                      \
+		val = (++val < max) ? val : 0;                                 \
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:22]: optional: this is hard to read.  How about

val = (val == max - 1) ? 0 : val + 1

File: drivers/i2s/i2s_litex.c [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+	}
+
+/**
+ * @brief Enable i2s device
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_enable(int reg)
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:32]: Consider using uintptr_t instead of int to make it clearer this is an address.

File: drivers/i2s/i2s_litex.c [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+}
+
+/**
+ * @brief Get i2s format handled by device 
+ *
+ * @param reg base register of device 
+ *
+ * @return currently supported format
+ */
+static int i2s_get_foramt(int reg)
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:74]: optional: returning an 'enum i2s_fmt' (if there is such a thing) here makes the types easier to debug.

File: drivers/i2s/i2s_litex.c [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+
+	litex_write8(reg_data | I2S_FIFO_RESET, reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Get i2s format handled by device 
+ *
+ * @param reg base register of device 
+ *
+ * @return currently supported format
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:72]: It also returns an error on error.

File: drivers/i2s/i2s_litex.c [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_clear_pending_irq(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_EV_PENDING_REG_OFFSET);
+
+	sys_write8(reg_data, reg + I2S_EV_PENDING_REG_OFFSET);
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:159]: Why sys_write8 instead of litex_write8?

File: drivers/i2s/i2s_litex.c [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+		data = sys_read32(I2S_RX_FIFO_ADDR);
+		for (int off = max_off; off >= 0; off--) {
+#if CONFIG_I2S_LITEX_DATA_BIG_ENDIAN
+			*(dst + i * chan_size + (max_off - off)) = data >> 8 * off;
+#else
+			*(dst + i * chan_size + off) = data >> 8 * off;
+#endif
+
+		}
+		// if mono, copy every left channel
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:204]: Use C style instead of C++ style comments.

File: drivers/i2s/i2s_litex.c [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+		for (size_t i = 0; i < size / chan_size; i += 2) {
+			data = sys_read32(I2S_RX_FIFO_ADDR);
+			*((uint16_t*)(dst + i)) = data & 0xffff;
+		}
+	}
+#else
+	int max_off = chan_size - 1;
+	for (size_t i = 0; i < size / chan_size; ++i) {
+		data = sys_read32(I2S_RX_FIFO_ADDR);
+		for (int off = max_off; off >= 0; off--) {
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:196]: I'm not certain that Zephyr style allows defining variables in the for loop initialiser.  Lets see what others say.

File: drivers/i2s/i2s_litex.c [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+}
+
+static int i2s_litex_configure(struct device *dev, enum i2s_dir dir,
+			       struct i2s_config *i2s_cfg)
+{
+	struct i2s_litex_data *const dev_data = DEV_DATA(dev);
+	const struct i2s_litex_cfg *const cfg = DEV_CFG(dev);
+	struct stream *stream;
+	int channels_concatenated;
+	if (dir == I2S_DIR_RX) {
```
Comment (nzmichaelh) [inline drivers/i2s/i2s_litex.c:332]: Add a space between the variable definition and the body.

File: dts/riscv/riscv32-litex-vexriscv.dtsi [hunk 11]
```diff
@@ -61,7 +61,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 		};
-		timer0: serial@e0002800 {
+		timer0: timer@e0002800 {
```
Comment (nzmichaelh) [inline dts/riscv/riscv32-litex-vexriscv.dtsi:64]: nit: this is an unrelated change so please split out into a different PR.

File: dts/riscv/riscv32-litex-vexriscv.dtsi [hunk 12]
```diff
@@ -137,5 +137,29 @@
...
 			#pwm-cells = <2>;
 		};
+		i2s_rx: i2s_rx@e000a800 {
+			compatible = "litex,i2s";
+			reg = <0xe000a800 0x20 0xb1000000 0x40000>;
+			interrupt-parent = <&intc0>;
+			interrupts = <6 2>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg-names = "control","fifo";
```
Comment (nzmichaelh) [inline dts/riscv/riscv32-litex-vexriscv.dtsi:147]: Add a space between values

**Review by nzmichaelh — DISMISSED** [review]

File: drivers/i2s/i2s_litex.c (hunk 5 above)
Comment (pawelsag) [reply drivers/i2s/i2s_litex.c:74]: I want to be compatible with i2s standard header and reuse defined variables.
Unfortunately, they are defined as #define. The type for these values comes with i2s header.
I will change returned type to i2s_fmt_t.

File: dts/riscv/riscv32-litex-vexriscv.dtsi (hunk 11 above)
Comment (mateusz-holenko) [reply dts/riscv/riscv32-litex-vexriscv.dtsi:64]: You are right, this is not related to I2S - that's why it's put into a separate commit. Is there a point of creating a separate PR just for this small change? If so, we'll ofc split this.

File: dts/riscv/riscv32-litex-vexriscv.dtsi (hunk 12 above)
Comment (mateusz-holenko) [reply dts/riscv/riscv32-litex-vexriscv.dtsi]: fixed

File: drivers/i2s/i2s_litex.c (hunk 10 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c:342]: added

File: drivers/i2s/i2s_litex.c (hunk 8 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: changed style of all comments

File: drivers/i2s/i2s_litex.c (hunk 7 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: For a single byte there is no difference between the two, but I agree that for consistency it should be `litex_write8`. It's fixed now.

File: drivers/i2s/i2s_litex.c (hunk 6 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: Updated the comment

File: drivers/i2s/i2s_litex.c (hunk 4 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: Changed here and in simillar functions

File: drivers/i2s/i2s_litex.c (hunk 3 above)
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: Refactored as suggested

File: drivers/i2s/Kconfig.litex (hunk 2 above)
Comment (mateusz-holenko) [reply drivers/i2s/Kconfig.litex]: Fixed

File: drivers/i2s/Kconfig.litex (hunk 1 above)
Comment (mateusz-holenko) [reply drivers/i2s/Kconfig.litex]: Changed to `n`

**Comment by mateusz-holenko** [comment]
@nzmichaelh Implemented your remarks, could you take look at the current version?

**Comment by mateusz-holenko** [comment]
@nzmichaelh Is there anything else we should improve in this PR?

**Review by kgugala — CHANGES_REQUESTED** [review]
One small comment, otherwise LGTM

File: dts/riscv/riscv32-litex-vexriscv.dtsi (hunk 11 above)
Comment (mateusz-holenko) [reply dts/riscv/riscv32-litex-vexriscv.dtsi:64]: I extracted this commit to a separate PR: #27519

File: drivers/i2s/i2s_litex.c [hunk 13]
```diff
@@ -0,0 +1,646 @@
...
+ *
+ * @return currently supported format or error
+ *			when format can't be handled
+ */
+static i2s_fmt_t i2s_get_foramt(uintptr_t reg)
+{
+	uint8_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+
+	reg_data &= I2S_CONF_FORMAT_MASK;
+	if (reg_data == 1) {
```
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply drivers/i2s/i2s_litex.c]: it's fixed now

**Review by kgugala — APPROVED** [review]

**Review by nashif — APPROVED** [review]
